### ISSUE TITLE
.pnpmfile change not triiger resolution

### ISSUE THREAD

**Issue by tjx666** [pnpm/pnpm#6726]
### Code to reproduce the issue:

https://stackblitz.com/edit/node-3neeb1?file=.pnpmfile.cjs

1. pnpm install
2. change log statement in `.pnpmfile.cjs`
3. pnpm install, you will see `Lockfile is up to date, resolution step is skipped`

### Expected behavior:

change `.pnpmfile.cjs` should always trigger resolution

### Actual behavior:

as title

### Additional information:

 - `node -v` prints: v16.20.0
 - Windows, macOS, or Linux?: https://stackblitz.com/

**Comment by zkochan** [comment]
How can we know if the file has changed? We could store the hash of the file in the lockfile. However, the hash won't change if the will happen not in `.pnpmfile.cjs` itself, but in modules that it requires.

**Comment by tjx666** [comment]
at least change .pnpmfile.cjs should trigger resolution

**Comment by thynson** [comment]
Let me try to understand the point of @tjx666 . The pnpmfile file affects resolution, since it can mutate the dependencies of a package in hooks, so it's reasonable that any modification to it need to trigger an resolution. The best way of it can be store the hash of pnpmfile in the lockfile, and possibly the file path of it, if renaming a pnpmfile, e.g., from pnpmfile.js to pnpmfile.js hit a corner case.

**Comment by icy0307** [comment]
https://github.com/pnpm/pnpm/issues/4794
this is the same issue.
Every time someone modifies package.json first,  the hooks on `master` branch become useless, until the next time someone installs or removes something in package.json
Can we do some ast parsing recursively and find all related files?  Store the hash for pnpmfile and related "required" files.

**Comment by thynson** [comment]
> Let me try to understand the point of @tjx666 . The pnpmfile file affects resolution, since it can mutate the dependencies of a package in hooks, so it's reasonable that any modification to it need to trigger an resolution. The best way of it can be store the hash of pnpmfile in the lockfile, and possibly the file path of it, if renaming a pnpmfile, e.g., from pnpmfile.js to pnpmfile.js hit a corner case.

I realized storing the hash value of pnpmfile.js in the lockfile is a bad idea. Instead, make sure the hooks in the pnpmfile.js are invoked for each dependencies in package.json, before checking a lockfile is up-to-date.

**Comment by icy0307** [comment]
But doesn't the readPackage hook get called during the resolution step?
@thynson

**Comment by thynson** [comment]
@icy0307 It seems like not for every dependencies, according to https://github.com/pnpm/pnpm/issues/4794#issuecomment-1331856002.

**Comment by icy0307** [comment]
@thynson  yes, not invoke for every dependency only if  resolution get skipped.  
And this because current hooks get called in resolution step.

**Comment by lake2** [comment]
> Fixed by [#7662](https://github.com/pnpm/pnpm/pull/7662)

This feature behaves different  between v9 and v10.
In v9, `readPackage` be called everytime when I execute  `pnpm i` continuously.
In v10, `readPackage` be called onetime when I execute  `pnpm i` continuously.
